evap/staff/importers/json.py

```diff
@@ -0,0 +1,325 @@
+    attempted_changes: list[Evaluation] = field(default_factory=list)
+
+    @staticmethod
+    def _make_heading(heading: str) -> str:
+        heading += "\n" + "".join(["-" for i in heading]) + "\n"
+        return heading
+
+    @staticmethod
+    def _make_total(total: int) -> str:
+        return f"({total} in total)\n\n"
+
+    def get_log(self) -> str:
+        log = "JSON IMPORTER REPORT\n"
+        log += "====================\n\n"
+        log += f"Import finished at {now()}\n\n"
+        log += self._make_heading("Name Changes")
+        for name_change in self.name_changes:
+            log += f"- {name_change.old_first_name_given} {name_change.old_last_name} → {name_change.new_first_name_given} {name_change.new_last_name}\n"
+        log += self._make_total(len(self.name_changes))
+
+        log += self._make_heading("New Courses")
+        for new_course in self.new_courses:
+            log += f"- {new_course}\n"
+        log += self._make_total(len(self.new_courses))
+
+        log += self._make_heading("New Evaluations")
+        for new_evaluation in self.new_evaluations:
+            log += f"- {new_evaluation}\n"
+        log += self._make_total(len(self.new_evaluations))
+
+        log += self._make_heading("Updated Courses")
+        for updated_course in self.updated_courses:
+            log += f"- {updated_course}\n"
+        log += self._make_total(len(self.updated_courses))
+
+        log += self._make_heading("Updated Evaluations")
+        for updated_evaluation in self.updated_evaluations:
+            log += f"- {updated_evaluation}\n"
+        log += self._make_total(len(self.updated_evaluations))
+
+        log += self._make_heading("Attempted Changes")
+        for attempted_change in self.attempted_changes:
+            log += f"- {attempted_change}\n"
+        log += self._make_total(len(self.attempted_changes))
+
+        return log
+
+
+class JSONImporter:
+    DATETIME_FORMAT = "%d.%m.%Y %H:%M"
+
+    def __init__(self, semester: Semester) -> None:
+        self.semester = semester
+        self.user_profile_map: dict[str, UserProfile] = {}
+        self.course_type_cache: dict[str, CourseType] = {}
+        self.degree_cache: dict[str, Degree] = {}
+        self.course_map: dict[str, Course] = {}
+        self.statistics = ImportStatistics()
+
+    def _get_course_type(self, name: str) -> CourseType:
+        if name in self.course_type_cache:
+            return self.course_type_cache[name]
+
+        course_type = CourseType.objects.get_or_create(name_de=name, defaults={"name_en": name})[0]
+        self.course_type_cache[name] = course_type
+        return course_type
+
+    def _get_degree(self, name: str) -> Degree:
+        if name in self.degree_cache:
+            return self.degree_cache[name]
+
+        degree = Degree.objects.get_or_create(name_de=name, defaults={"name_en": name})[0]
+        self.degree_cache[name] = degree
+        return degree
```

Comment:
These are duplicated with the user importer, right? Do we want to factor this out into a separate component?

Reply:
The mapper in the user importer expects objects to exist whether I want to create non-existing objects.